Add conditions for sidebarAction / sidePanel to WebExtensionAPINamespace::isPropertyAllowed.

AriYoung00 opened the pull request:
#### a0ee10281281011863adade244a37ff6dc0da272
<pre>
Add conditions for sidebarAction / sidePanel to WebExtensionAPINamespace::isPropertyAllowed.
<a href="https://webkit.org/b/277298">https://webkit.org/b/277298</a>
<a href="https://rdar.apple.com/132763372">rdar://132763372</a>

Reviewed by Timothy Hatcher and Brian Weinstein.

This patch adds the correct conditions for sidebarAction and sidePanel
to WebExtensionAPINamespace::isPropertyAllowed. It also adds the
sidePanel permission to the list of recognized valid permissions, so
that we can correctly parse it from the manifest. Finally, it also adds
a set of unit tests that exercise the added conditions for property
visibility.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.h: Declare
  _WKWebExtensionPermissionSidePanel
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.mm: Define
  _WKWebExtensionPermissionSidePanel
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::supportedPermissions): Add
_WKWebExtensionPermissionSidePanel to the list of supported permissions.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed): Add correct
conditions for sidePanel / sidebarAction properties
* Tools/TestWebKitAPI/SourcesCocoa.txt: Add WKWebExtensionAPISidebar.mm
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj: Add
  WKWebExtensionAPISidebar.mm
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm: Added.
(TestWebKitAPI::WKWebExtensionAPISidebarTest::WKWebExtensionAPISidebarTest): Add test fixture for sidebar tests
(TestWebKitAPI::TEST_F): Add various tests exercising property
conditions / argument count validation.

Canonical link: <a href="https://commits.webkit.org/281585@main">https://commits.webkit.org/281585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f5865c656d32fe7b950e71699b6744269950154

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39760 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12967 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/64328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/10940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62539 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11173 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/64328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/10940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62440 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37042 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/52335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/64328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/33740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/9560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/9857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/9848 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/66060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4342 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/66060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4361 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52309 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/66060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/3612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9068 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/35570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/37741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

AriYoung00 commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.h:
```diff
@@ -75,6 +75,10 @@ WK_EXTERN _WKWebExtensionPermission const _WKWebExtensionPermissionNativeMessagi
 
+/*! @abstract The `sidePanel` permission requests access to the `chrome.sidePanel` APIs. */
+WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))
+WK_EXTERN _WKWebExtensionPermission const _WKWebExtensionPermissionSidePanel;
```
Note: I wanted to condition this behind `ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)`, but the build system wouldn't allow me to include `wtf/Platform.h` in this file. Does anyone know if there's a workaround for this?

b-weinstein replied:
My understanding is that we don't want to have `#if` statements in API header files - since third party apps will import those and they don't know what they mean. I'm not sure what this means for the best way to handle this though - maybe to always include them? @xeenon might have a better idea on the best way ti handle it

xeenon replied:
We should put it in _WKWebExtensionPermissionsPrivate.h until it is ready to ship and has gone through API review.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -2103,7 +2103,11 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+        _WKWebExtensionPermissionStorage, _WKWebExtensionPermissionTabs, _WKWebExtensionPermissionUnlimitedStorage, _WKWebExtensionPermissionWebNavigation, _WKWebExtensionPermissionWebRequest,
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+        _WKWebExtensionPermissionSidePanel,
+#endif // ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
```
Super-nit: For one line blocks like this, we can probably omit the comment on the `#endif`.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -75,14 +75,12 @@
-        return false;
-
-    // FIXME: <https://webkit.org/b/276833> Check if the sidebar feature flag is enabled and if we have a sidebarAction key in the manifest to determine if this property should be visible
+        return page->corePage()->settings().webExtensionSidebarEnabled() && extensionContext().hasPermission("sidePanel"_s) && !objectForKey<NSDictionary>(extensionContext().manifest(), @"sidebar_action", true);
```
Just for my notes - what is the `true` argument that's the last one passed to `objectForKey`? It looks like other callers of that method use `false` here.

xeenon replied:
I think it is nil if empty when true, but I'm not looking at the header currently either. We should make these use an enum for this reason!

AriYoung00 replied:
Yep, the parameter is `nilIfEmpty`, which I think is more appropriate here since there is at least one sub-key to `sidebar_action` which is mandatory.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -0,0 +1,243 @@
+
+#import <WebKit/WKPreferencesPrivate.h>
+#import <WebKit/_WKFeature.h>
+
```
Nit: Extra blank space here.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -0,0 +1,243 @@
+};
+
+// This test fixture allows us to use sidebarConfig (which enables the sidebar feature flag) without manually constructing one on each run
+class WKWebExtensionAPISidebarTest : public testing::Test {
```
Neat! Do we do this anywhere else?

AriYoung00 replied:
I haven't seen it in the other extensions tests, but it looks like it's used extensively in the WebCore, WTF, and IPC tests

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -0,0 +1,243 @@
+TEST_F(WKWebExtensionAPISidebarTest, SidebarActionAPIAvailableWhenManifestRequests)
+{
+    auto *script = @[
+        @"browser.test.assertFalse(browser.sidebarAction === undefined)",
```
I wish there was a better way to say something is **not** undefined.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -0,0 +1,243 @@
+        @"browser.test.assertDeepEq(browser.sidebarAction, undefined)",
+        @"browser.test.assertDeepEq(browser.sidePanel, undefined)",
+        @"browser.test.assertDeepEq(chrome.sidebarAction, undefined)",
+        @"browser.test.assertDeepEq(chrome.sidePanel, undefined)",
```
Nit: I'm not sure we need the tests of the `chrome` namespace. We alias it to browser everywhere and I believe that is tested elsewhere.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -75,14 +75,12 @@
 
 #if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
-    // FIXME: <https://webkit.org/b/276833> Check if the sidebar feature flag is enabled and if we have the sidePanel permission to determine if this property should be visible
+    // If the extension requests both sidePanel and sidebarAction, we will give them sidebarAction -- we check in sidePanel that there is no sidebar_action key, but we do not check in sidebarAction that there is no sidePanel permission
```
This could be multiple lines since it is a little long.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -0,0 +1,243 @@
+};
+
+// This test fixture allows us to use sidebarConfig (which enables the sidebar feature flag) without manually constructing one on each run
+class WKWebExtensionAPISidebarTest : public testing::Test {
```
This should be `WKWebExtensionAPISidebar` (drop the Test suffix.) that way it is consistent with the other extension test categories.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.h:
```diff
@@ -75,6 +75,10 @@ WK_EXTERN _WKWebExtensionPermission const _WKWebExtensionPermissionNativeMessagi
 
+/*! @abstract The `sidePanel` permission requests access to the `chrome.sidePanel` APIs. */
+WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))
+WK_EXTERN _WKWebExtensionPermission const _WKWebExtensionPermissionSidePanel;
```
Move to `…Private.h`.

b-weinstein replied:
Ah that was it!

AriYoung00 replied:
Unfortunately, 'Check For Inappropriate Macros in External Headers' is still failing if I use `ENABLE` in `_WKWebExtensionPermissionPrivate.h`. Should I just leave this symbol visible in all cases?

kiaraarose replied:
i think that's fine

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.h:
```diff
@@ -75,6 +75,10 @@ WK_EXTERN _WKWebExtensionPermission const _WKWebExtensionPermissionNativeMessagi
 
+/*! @abstract The `sidePanel` permission requests access to the `chrome.sidePanel` APIs. */
+WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))
+WK_EXTERN _WKWebExtensionPermission const _WKWebExtensionPermissionSidePanel;
```
wouldn't we also need a permission for `sidebar_action`?

AriYoung00 replied:
I don't believe so -- the behavior in Firefox/Opera is that extensions are always allowed to use `sidebarAction` as long as their manifest specifies a `sidebar_action` key. Would I need to add anything to parse this? In my unit tests, it does seem to be visible in the manifest dictionary without any changes.

kiaraarose replied:
i see. so there's no actual `sidebar_action` permission, just the entry in the manifest?

AriYoung00 replied:
Yep, exactly

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.h:
```diff
@@ -75,6 +75,10 @@ WK_EXTERN _WKWebExtensionPermission const _WKWebExtensionPermissionNativeMessagi
 WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
 WK_EXTERN _WKWebExtensionPermission const _WKWebExtensionPermissionScripting;
 
+/*! @abstract The `sidePanel` permission requests access to the `chrome.sidePanel` APIs. */
```
Even though this is a Chrome specific API currently, we should use `browser` to be neutral.

```suggestion
/*! @abstract The `sidePanel` permission requests access to the `browser.sidePanel` APIs. */
```

AriYoung00 committed:
Add conditions for sidebarAction / sidePanel to WebExtensionAPINamespace::isPropertyAllowed.
https://webkit.org/b/277298
rdar://132763372

Reviewed by Timothy Hatcher and Brian Weinstein.

This patch adds the correct conditions for sidebarAction and sidePanel
to WebExtensionAPINamespace::isPropertyAllowed. It also adds the
sidePanel permission to the list of recognized valid permissions, so
that we can correctly parse it from the manifest. Finally, it also adds
a set of unit tests that exercise the added conditions for property
visibility.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.h: Declare
  _WKWebExtensionPermissionSidePanel
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.mm: Define
  _WKWebExtensionPermissionSidePanel
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::supportedPermissions): Add
_WKWebExtensionPermissionSidePanel to the list of supported permissions.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed): Add correct
conditions for sidePanel / sidebarAction properties
* Tools/TestWebKitAPI/SourcesCocoa.txt: Add WKWebExtensionAPISidebar.mm
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj: Add
  WKWebExtensionAPISidebar.mm
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm: Added.
(TestWebKitAPI::WKWebExtensionAPISidebarTest::WKWebExtensionAPISidebarTest): Add test fixture for sidebar tests
(TestWebKitAPI::TEST_F): Add various tests exercising property
conditions / argument count validation.

Canonical link: https://commits.webkit.org/281585@main
